### ISSUE TITLE
How to work with multiple client Sessions?

### ISSUE THREAD

**Issue by Potherca** [inrupt/solid-client-authn-js#1685]
I am using `@inrupt/solid-client-authn-browser` and I am trying to have two separate sessions on one page.

The idea is that a user can see the contents of different pods on the same page.

However, I am running into issues `handleIncomingRedirect()` overwriting the _both_ sessions, not just the one it is attached to... 

```js
  const sessions = {
    source: new Session({}),
    target: new Session({}),
  }

  for (const key in sessions) {
    const session = sessions[key];

    session.handleIncomingRedirect({url: `${window.location.href}`, restorePreviousSession: true})
  }
```

I tried manually setting the `sessionId`, adding the sessionId as a GET param to the `url`, and I've fiddled with `restorePreviousSession` but I just can seem to get this library to only change the data for _one_ session.

Can anybody point me into a direction that might work?

**Comment by NSeydoux** [comment]
Maybe a first note (in order to make sure that we are trying to solve the right problem) is that nothing prevents a single identity to interact with several Pods. A user may log in once, and given the obtained session read/write data into as many Pods as they want, provided they have been given the appropriate access level. 

The problem with what you are trying to achieve is that by essence, one redirect makes it possible to log in one session. That's how OIDC works, so we that's not something we can change. One thing that may be considered is for the user to log in their first identity, get redirected back to the app, get the first set of information displayed, and then log into their second identity, get redirected again, and then see the second set of information. 

Another issue is that if you're working purely in the browser, redirecting the user away from the page loses the session information, including the access token. So logging in the second identity would necessarily log the first one out. This constraint could be worked around by introducing a server component, and using `@inrupt/solid-client-authn-node` instead.

I guess the real question is: is there any way what you intend to achieve could work with the user logging in only once, and the obtained identity accessing the `source` and `target` Pods ?

**Comment by Potherca** [comment]
Thank you for pointing out OIDC. I _should_ have thought about myself, to be honest.
I guess my brain is still too hung up on the old paradigm of single user -> single login -> single storage.

However, I think you are correct about only needing one _session_. :thinking: 

My current setup was:

1. Log the user in
2. Grab the `pim/space#storage` from the WebID the user logged in with
3. Display contents from the Pod at the `pim/space#storage` URL

I was trying to do go through the same for the second Pod... But I don't need another session for that... Just an URL where the second Pod lives. I can indeed just re-use the session, as the _user_ hasn't changed :facepalm:.

Worst case scenario, if the user has two pods which use different WebIds, they will need to give one of their WebId access rights on the other Pod.

I guess that, if I _really_ need to support multiple sessions, I'll have to start using  iframes and/or popup windows. But I think I should be fine for the time being with just using the one session to make calls to multiple pods.

Thank you very much @NSeydoux!

**Comment by NSeydoux** [comment]
No problem, working in open ecosystem where identity, applications and storage are not tightly coupled is far from being obvious :D. Happy to help, and feel free to ask additional questions if you run into other issues, it also creates interesting references for other developers who run into the same problems. A final note: the `pim:storage` is not guaranteed to be present in a user's WebID (even if it is recommanded that it is), not that it is unique (you may even encourage users of your app to reference several storage from their WebIDs depending on the service you want to provide).

Cheers !
